rtic-sync/src/signal.rs

```diff
@@ -0,0 +1,119 @@
+//! A "latest only" value store with unlimited writers and async waiting.
+
+use core::{
+    cell::UnsafeCell,
+    future::poll_fn,
+    sync::atomic::{fence, Ordering},
+    task::Poll,
+};
+use rtic_common::waker_registration::CriticalSectionWakerRegistration;
+
+/// Basically an Option but for indicating
+/// whether the store has been set or not
+#[derive(Clone, Copy)]
+enum Store<T> {
+    Set(T),
+    Unset,
+}
+
+/// A "latest only" value store with unlimited writers and async waiting.
+pub struct Signal<T: Copy> {
+    waker: CriticalSectionWakerRegistration,
+    store: UnsafeCell<Store<T>>,
+}
+
+unsafe impl<T: Copy> Send for Signal<T> {}
+unsafe impl<T: Copy> Sync for Signal<T> {}
+
+impl<T: Copy> Signal<T> {
+    /// Create a new signal.
+    pub const fn new() -> Self {
+        Self {
+            waker: CriticalSectionWakerRegistration::new(),
+            store: UnsafeCell::new(Store::Unset),
+        }
+    }
+
+    /// Split the signal into a writer and reader.
+    pub fn split(&'static self) -> (SignalWriter<T>, SignalReader<T>) {
+        (SignalWriter { parent: self }, SignalReader { parent: self })
+    }
+}
+
+/// Fascilitates the writing of values to a Signal.
+#[derive(Clone)]
+pub struct SignalWriter<T: 'static + Copy> {
+    parent: &'static Signal<T>,
+}
+
+impl<T: Copy> SignalWriter<T> {
+    /// Write a raw Store value to the Signal.
+    fn write_inner(&mut self, value: Store<T>) {
+        fence(Ordering::SeqCst);
+
+        critical_section::with(|_| {
+            // SAFETY: in a cs: exclusive access
+            unsafe { self.parent.store.get().replace(value) };
+        });
+
+        self.parent.waker.wake();
+    }
+
+    /// Write a value to the Signal.
+    pub fn write(&mut self, value: T) {
+        self.write_inner(Store::Set(value));
+    }
+
+    /// Clear the stored value in the Signal (if any).
+    pub fn clear(&mut self) {
+        self.write_inner(Store::Unset);
+    }
+}
+
+/// Fascilitates the async reading of values from the Signal.
+pub struct SignalReader<T: 'static + Copy> {
+    parent: &'static Signal<T>,
+}
+
+impl<T: Copy> SignalReader<T> {
+    /// Immediately read and evict the latest value stored in the Signal.
+    fn take(&mut self) -> Store<T> {
+        critical_section::with(|_| {
+            // SAFETY: in a cs: exclusive access
+            unsafe { self.parent.store.get().replace(Store::Unset) }
+        })
+    }
+
+    /// Wait for a new value to be written and read it.
+    ///
+    /// If a value is already pending it will be returned immediately.
+    pub async fn wait(&mut self) -> T {
+        poll_fn(|ctx| match self.take() {
+            Store::Unset => {
+                self.parent.waker.register(ctx.waker());
```

Comment:
This is a race-condition.

The registration must happen before `self.take()` on line 91 as the value can be set (and fail to be woken) between the `self.take()` but before the waker registration.

Reply:
Ah very interesting. I was trying to avoid registration if a value is already there, but it seems unavoidable.

By the way, I noticed the WakerRegistration never *clears* the registered waker, and during my experimentation I noticed that calling wake on a Waker belonging to an expired Future didn't do anything. I assume this is a feature of the rtic executor but was curious if this was truly the case?

Reply:
Yeah it's generally unavoidable - but the async executor and tasks are robust to spurious wakes. So in case an extra wake does happen it's no problem. 

One can add an API to the waker registration to clear a registration if it was not needed, but I've never measured the impact of it.
And if we don't measure we don't know if it will be better or worse :)

Reply:
I actually did add that at one point but removed it because it did incur a slight performance hit, and of course I observed no issue with *not* clearing it.